src/main/java/org/folio/rest/impl/SamlAPI.java

```diff
@@ -579,4 +609,48 @@ private SamlConfig configToDto(SamlConfiguration config) {
     return samlConfig;
   }
 
+  @Override
+  public void optionsSamlLogin(RoutingContext routingContext, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    handleOptions(routingContext);
+  }
+
+  @Override
+  public void optionsSamlCallback(RoutingContext routingContext, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    handleOptions(routingContext);
+  }
+
+  private void handleOptions(RoutingContext routingContext) {
+    HttpServerRequest request = routingContext.request();
+    HttpServerResponse response = routingContext.response();
+    String origin = request.headers().get(ORIGIN);
+    if (origin == null) {
+      response.setStatusCode(400).setStatusMessage("Missing origin header").end();
+      return;
+    }
+    if (origin.isBlank() || origin.trim().contentEquals("*")) {
+      response.setStatusCode(400).setStatusMessage("Invalid origin header").end();
+      return;
+    }
+    response.putHeader(ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+    response.putHeader(ACCESS_CONTROL_ALLOW_ORIGIN, origin);
+    Utils.appendToMapIfAbsent(response.headers(), VARY, ",", ORIGIN);
+    response.putHeader(ACCESS_CONTROL_ALLOW_METHODS, request.getHeader(ACCESS_CONTROL_REQUEST_METHOD));
+    Utils.appendToMapIfAbsent(response.headers(), VARY, ",", ACCESS_CONTROL_REQUEST_METHOD);
+    response.putHeader(ACCESS_CONTROL_ALLOW_HEADERS, request.getHeader(ACCESS_CONTROL_REQUEST_HEADERS));
+    Utils.appendToMapIfAbsent(response.headers(), VARY, ",", ACCESS_CONTROL_REQUEST_HEADERS);
+    response.setStatusCode(204).end();
+  }
+
+  private void addCredentialsAndOriginHeaders(RoutingContext routingContext) {
+    String origin = routingContext.request().headers().get(ORIGIN);
+    if (origin == null || origin.isBlank() || origin.trim().contentEquals("*")) {
```

Comment:
line 648 is the same as line 632+628 and might be extracted into `isInvalidOrigin(String)`.
line 632+628 can be combined, null is also invalid.

Reply:
> line 648 is the same as line 632+628 and might be extracted into `isInvalidOrigin(String)`.
> line 632+628 can be combined, null is also invalid.

Done.